### ISSUE TITLE
Improve sheets building process

### ISSUE THREAD

**Issue by exerro** [Exerro/Sheets#39]
The current way sheets builds is... horrible. Ideally, the final version would output a single file with everything localised, and a lookup table for file lines that's used by the Exception class to correctly determine line numbers and source files.

The preprocessor could be used to make `@class XYZ` turn into `local XYZ = class "XYZ"`, and localise all classes at the top of the build.

**Comment by exerro** [comment]
Line-preserving minification would be very good too.

**Comment by CrazedProgrammer** [comment]
You could do LZW-compression and base92 packing like carl

**Comment by exerro** [comment]
Wouldn't it then need to ship with a decompresser? Seems a little too complicated tbh. Also why thte base92 packing?

**Comment by CrazedProgrammer** [comment]
Yes it ships with a ~1kb decompresser, it uses base92 packing because you need to inline it in a lua string.

**Comment by exerro** [comment]
Seems somewhat verbose. Maybe in a later version. Also rip program startup time.

**Comment by CrazedProgrammer** [comment]
It's actually not noticeable.
It takes less than 1/10th of a second to decompress Surface 2 from 27kB to 39kB.

**Comment by exerro** [comment]
Sheets is ~120kb without surface, ~160kb with surface, and will probably be more like 200kb in total by the time v0.0.4 is released. Assuming it takes 5x longer (as it's 5x the uncompressed size), that's half a second of delay. It's not much, sure, but still a pain. Half a second is actually really noticeable.

**Comment by CrazedProgrammer** [comment]
Yeah you have a point.

**Comment by exerro** [comment]
The new build system implemented in 324eab0 changed how classes work, most importantly the fact that they are all now locals rather than in a sandboxed environment. This means code generation that relied on accessing these classes now fails as those classes aren't in the environment passed to them. While a patch has been applied (see [Codegen.lua](https://github.com/Exerro/Sheets/blob/develop/dynamic/Codegen.lua#L309)), there might still be issues. It needs to be reviewed.
